lib/qbo_api.rb

```diff
@@ -112,26 +113,37 @@ def request(method, path:, entity: nil, payload: nil, params: nil)
   end
 
   def response(resp, entity: nil)
-    data = JSON.parse(resp.body)
-    if entity
-      entity_response(data, entity)
-    else
-      data
-    end
+    data = parse_response_body(resp)
+    entity ? entity_response(data, entity) : data
   rescue => e
-    # Catch fetch key errors and just return JSON
+    QboApi.logger.debug { "#{LOG_TAG} response parsing error: entity=#{entity.inspect} body=#{resp.body.inspect} exception=#{e.inspect}" }
```

Comment:
One last thing. This one is a bit noisy on specs e.g.

```
bundle exec rspec spec/
D, [2018-03-20T19:11:41.754984 #15262] DEBUG -- : [QuickBooks] entity name not in response body: entity=:customer entity_name="Customer" body={"QueryResponse"=>{"totalCount"=>29}, "time"=>"2015-11-05T04:30:59.195-08:00"}
.D, [2018-03-20T19:11:41.771654 #15262] DEBUG -- : [QuickBooks] entity name not in response body: entity=:employee entity_name="Employee" body={"QueryResponse"=>{"totalCount"=>2}, "time"=>"2016-10-04T04:03:53.319-07:00"}
.D, [2018-03-20T19:11:41.797480 #15262] DEBUG -- : [QuickBooks] entity name not in response body: entity=:vendor entity_name="Vendor" body={"QueryResponse"=>{"totalCount"=>26}, "time"=>"2015-11-05T06:27:43.230-08:00"}
.D, [2018-03-20T19:11:41.854119 #15262] DEBUG -- : [QuickBooks] entity name not in response body: entity=:customer entity_name="Customer" body={"QueryResponse"=>{"totalCount"=>6}, "time"=>"2016-10-04T04:07:15.826-07:00"}
..D, [2018-03-20T19:11:41.907760 #15262] DEBUG -- : [QuickBooks] entity name not in response body: entity=:customer entity_name="Customer" body={"QueryResponse"=>{"totalCount"=>29}, "time"=>"2015-11-05T04:30:59.195-08:00"}
..D, [2018-03-20T19:11:41.925671 #15262] DEBUG -- : [QuickBooks] entity name not in response body: entity=:attachable entity_name="Attachable" body={"AttachableResponse"=>[{"Fault"=>{"Error"=>[{"Message"=>"A business validation error has occurred while processing your request", "Detail"=>"Business Validation Error: The entity Estimate 75 that you are trying to link does not exist.", "code"=>"6000", "element"=>""}], "type"=>"ValidationFault"}}], "time"=>"2018-01-06T06:32:09.863-08:00"}
.............................................
Finished in 0.39656 seconds (files took 0.44675 seconds to load)
52 examples, 0 failures
```

Any ideas to keep things cleaner on runs?

Reply:
done in 21c057e